### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by lachrist**
Before, issues *probably* caused by the user were handled by exiting the process and issues *probably* caused by a bug in the agent were handled by throwing an error. The idea behind calling `process.exit` was to prevent the user code from catching and silencing appmap-related errors. This is bad because it can prevent the user from receiving feedbacks on how to use the agent correctly.

In this PR I unified issues handling and get rid of calls to `process.exit` which is a very crude mechanism. Issues *probably* caused by the user are handled by logging a message and by throwing an `ExternalAppmapError`. Even if the error is silenced, the log message should help the user diagnose the issue. And issues *probably* caused by the user are handled by throwing an `InternalAppmapError`.

TLDR:

- Calls to `process.exit` are replace by throw expressions.
- Error messages are short and static.
- Logs can contain dynamic data via formatting.
- `ExternalAppmapError` *probably* flags an issue caused by the user 
- `InternalAppmapError` *probably* flags a bug in the agent
- Other errors: either caused by the user code or caused by a bug in the agent.

I also added support for node 19 which only required test update.

### CODE REVIEW

**Review by dustinbyrne — CHANGES_REQUESTED** [review]
~I'm still working my way through this PR but I figured I'd submit the comments I've found so far.~

Everything else looks good! Please see the comments below.

File: components/error/default/index.mjs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Detected an external appmap error.\
+ This is probably a issue with how you used the appmap framework.
+- You can look for answers in our installation instructions:
+    ${instruction}
+- We also have a reference documentation:
+    ${documentation}
+- You can ask for help in our public slack channel:
+    ${slack}
+- We also have a public discord:
+    ${discord}
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/error/default/index.mjs]: Done in https://github.com/getappmap/appmap-agent-js/commit/afc879547c5d9705d8276f02f8330840e3d2c971

File: components/error/default/index.mjs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+const slack = "https://appmap-group.slack.com/archives/C03GEJVDC2J";
+const discord =
+  "https://discord.com/channels/766016904056930325/818522100205223996";
+const documentation =
+  "https://github.com/getappmap/appmap-agent-js/blob/main/REFERENCE.md";
+const instruction = "https://appmap.io/docs/appmap-overview.html";
+
+const internal_message = `
+Detected an internal appmap error.\
+ This is probably an issue within the appmap framework.\
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/error/default/index.mjs]: Done in https://github.com/getappmap/appmap-agent-js/commit/afc879547c5d9705d8276f02f8330840e3d2c971

File: components/util/default/assert.mjs [hunk 3]
```diff
@@ -1,18 +1,28 @@
...
   constructor(message) {
     super(message);
     this.name = "AssertionError";
   }
 }
 
-export const assert = (boolean, message) => {
+export const assert = (boolean, message, Constructor) => {
+  if (typeof Constructor !== "function") {
+    throw new Error("BOUM");
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/util/default/assert.mjs:14]: Done in https://github.com/getappmap/appmap-agent-js/commit/afc879547c5d9705d8276f02f8330840e3d2c971

File: components/error/default/index.mjs [hunk 4]
```diff
@@ -0,0 +1,68 @@
+const { URL, Error } = globalThis;
+
+const { search: __search } = new URL(import.meta.url);
+
+const issues = "https://github.com/getappmap/appmap-agent-js/issues";
+const slack = "https://appmap-group.slack.com/archives/C03GEJVDC2J";
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/error/default/index.mjs]: Done in https://github.com/getappmap/appmap-agent-js/commit/afc879547c5d9705d8276f02f8330840e3d2c971

File: components/error/default/index.mjs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+Detected an internal appmap error.\
+ This is probably an issue within the appmap framework.\
+ Please consider submitting a bug report at:
+  ${issues}
+
+`;
+
+const external_message = `
+Detected an external appmap error.\
+ This is probably a issue with how you used the appmap framework.
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/error/default/index.mjs]: Done in https://github.com/getappmap/appmap-agent-js/commit/afc879547c5d9705d8276f02f8330840e3d2c971

File: components/util/default/assert.mjs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lachrist) [reply components/util/default/assert.mjs]: This was added to make sure I updated all assert call. I forgot to remove it afterward :( Thanks for noticing! I rebased in https://github.com/getappmap/appmap-agent-js/commit/97132a8fdeafb89f827f5742099c329bc22302d0

File: components/util/default/assert.mjs [hunk 6]
```diff
@@ -1,18 +1,28 @@
...
-    throw new AssertionError(message);
+    throw new Constructor(message);
   }
 };
 
-export const generateDeadcode = (message) => () => {
-  throw new AssertionError(message);
+export const generateDeadcode = (message, Constructor) => {
+  if (typeof Constructor !== "function") {
+    throw new Error("BOUM");
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/util/default/assert.mjs]: Rebased in https://github.com/getappmap/appmap-agent-js/commit/97132a8fdeafb89f827f5742099c329bc22302d0

File: components/log/test/index.test.mjs [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+import {
+  logWarning,
+  logWarningWhen,
+  logInfo,
+} from "./index.mjs?env=test&log-level=warning";
+
+logInfo("foo");
+logWarning("bar");
+logWarningWhen(true, "qux");
+logWarningWhen(false, "qux");
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/log/test/index.test.mjs]: Yeah, it is not straightforward to check the message is written to the log file because the test version of `log-inner` is a noop. I thought the tested code was simple enough that I could skip it.

File: components/hook-cjs/void/index.mjs [hunk 8]
```diff
@@ -1,9 +1,19 @@
...
+);
+const { logErrorWhen } = await import(`../../log/index.mjs${__search}`);
+const { assert, noop } = await import(`../../util/index.mjs${__search}`);
+
+export const unhook = noop;
 
 export const hook = (_agent, { hooks: { cjs } }) => {
-  expect(!cjs, "expected configuration to disable cjs module hook");
+  assert(
+    !logErrorWhen(cjs, "No support for recordering common-js modules"),
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: Sorry for the typo: http://github.com/getappmap/appmap-agent-js/commit/6a889ae66ed072dccf02570e027c218c72f57884

[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: You got the condition in the log reversed. Here is the correct branching:
- If cjs is `truthy`, an error is *logged* and the assertion fails.
- If cjs is `falsy`, an error is *not logged* and the assertion passes.

[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: We could align the condition of the log on the assertion. But I don't think it is more readable:

```js
assert(
  logErrorWhenNot(!cjs, "No support for recording common-js modules"),
  "Expected hooks.cjs to be false",
  ExternalAppmapError,
);
```

File: components/hook-cjs/void/index.mjs [hunk 9]
```diff
@@ -1,9 +1,19 @@
...
+    "No support for recordering common-js modules",
```
[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: `assert` and `logErrorWhen` are complementary:
- `logErrorWhen`: log a message if the condition is true
  - Cannot be silenced by the user
  - Prints potentially long message containing dynamic data
- `assert`: throw an error if the condition is false
  - Can be caught and silenced by user code
  - Contain a short static message

In general I follow this pattern:
- If the error is probably *not* a bug in the agent I log an error message and throw an `ExternalAppmapError`.
- If the error is probably a bug in the agent I just throw an `InternalAgentError`.

[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: The code might be repetitive but it is explicit. With minimal changes we could have constructor that does the logging:

```js
throw new ExternalAppmapError(
  "short static message",
  "template with dynamic data: %j %j",
  123,
  456,
);

assert(
  condition,
  ExternalAppmapError,
  "short static message",
  "template with dynamic data: %j %j",
  123,
  456,
);
```

But I think I prefer the explicit version.

File: components/hook-cjs/void/index.mjs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply components/hook-cjs/void/index.mjs]: https://github.com/getappmap/appmap-agent-js/pull/138/files#diff-da1eca588b0fb0bee76240a2303f134eabe0901c16680407571783e4bc7165b5R114-R115
```js
logError("Could not recognize %j as a mocha command", tokens);
throw new ExternalAppmapError("Not a parsed mocha command");
```
It may just be me, but this is a lot more digestible as it's intent is clearly stated

**Review by dustinbyrne — APPROVED** [review]
Approved pending build results

File: components/hook-cjs/void/index.mjs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lachrist) [reply components/hook-cjs/void/index.mjs]: I agree, but they don't have a condition. So it would have to be:
```js
if (cjs) {
  logError("No support for recording common-js modules");
  throw new ExternalAppmapError("Expected hooks.cjs to be false");
}
```

Which is admittedly more readable. But it adds more branching to the code and I try to reach 100% coverage. Is this stupid? I don't want to add `/* c8 ignore */` around every assertion... Sometimes testability harms the code I guess?
